BUG: Travis building on container-based infrastructure

nparley opened the pull request:
- [ ] closes #10598
- [x] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

Travis is not always detecting sudo use and is running forks using container-based infrastructure. `JOB_NAME=27_slow_nnet_LOCALE` fails when using containers because of setlocale. This PR adds `sudo: required` to make sure travis does not use containers while sudo is needed in the travis set up scripts. 

Example: 
- https://travis-ci.org/nparley/pandas/jobs/124695152 (container)
- https://travis-ci.org/nparley/pandas/jobs/124778492 (non container)

nparley committed:
Use sudo: required to stop travis building on container-based infrastructure

jreback commented:
does this cause anything to fail (on all runs)? e.g. so it seems that we are NOT testing the locale stuff then ? (w/o the sudo)

nparley commented:
These are all the runs of a clean fork running on containers https://travis-ci.org/nparley/pandas/builds/124695150 (vs not https://travis-ci.org/nparley/pandas/builds/124778490). Most of the runs pass fine it seems. `27_nslow_nnet_COMPAT` also has the problem. `35_numpy_dev` also fails with:

```
ImportError: libatlas.so.3gf: cannot open shared object file: No such file or directory
```

jreback commented:
you may need to change the ``ci/install-3.5_NUMPY_DEV.sh`` to have that work, it does an ``sudo apt-et``......to install the system libraries (as its pulling down wheels, not conda envs for testing)

nparley commented:
Would the preference be to try and get travis working with containers over setting to sudo required?

jreback commented:
IIUC travis migrated us to a 'new' container infrastructure, but we are still doing it the 'old' way? is that accurate (IOW, the old warning has disappeared but they are doing some sort of magic to still put us on the containers). or is that completely wrong?

ideally we want to be doing this the most preferred way on travis. we don't care too much are actually using sudo as eveything lives in a conda env anyhow.

nparley commented:
From what I understand if you were using travis before 2015-01-01 you are still using standard infrastructure.

> * For repos we recognize before 2015-01-01, linux builds are sent to our standard infrastructure.
> * For repos we recognize on or after 2015-01-01, linux builds are sent to our container-based 

When you are running on containers you get the message:

```
This job is running on container-based infrastructure, which does not allow use of 'sudo', setuid and setguid executables.
If you require sudo, add 'sudo: required' to your .travis.yml
See https://docs.travis-ci.com/user/workers/container-based-infrastructure/ for details.
```

which does not appear for `pydata / pandas` builds. So I don't think pandas has been migrated. Or it was but now isn't again.

jreback commented:
``hostname: travis-worker-gce-org-prod4-4:8e9fb454-c5a1-4d27-bb67-cb0494ed1029``

so we are on gce. is that the container?

nparley commented:
No they are just ubuntu VMs on gce I think. It's a container if it has docker worker:

```
Using worker: worker-linux-docker-d4bd0a53.prod.travis-ci.org:travis-linux-14
```

jreback commented:
@nparley ok I guess we are not on containers. I *would* like to change. The main reason is that we can then use Travis own caching system, rather then our out-sourced version (with ``iron cache``), which while working is prob not as robust and integrated etc.

jreback commented:
@nparley does this now run on containers? can you rebase and repush

nparley commented:
@jreback no this PR actually is just forcing Travis *not* to run on containers to make sure it builds. I have just moved house the last two weeks so haven't had time to look at it again but will have a go at seeing if the locale stuff can be done at containers soon.

nparley committed:
Merge remote-tracking branch 'upstream/master' into travis

nparley committed:
First go to try locale settings on travis

nparley committed:
lanuage-pack-zh is not allowed try -chan

nparley committed:
Fix spelling

nparley committed:
Test zh_CN.UTF-8 as that's allowed to be installed

nparley committed:
Remove more sudo

jreback commented on ci/install-3.5_NUMPY_DEV.sh:
```diff
@@ -12,7 +12,8 @@ pip uninstall numpy -y
 # time conda install -n pandas libgfortran openblas || exit 1
 
-time sudo apt-get $APT_ARGS install libatlas-base-dev gfortran
+# Not going to work in containers
```
when you get this to work, you can just delete these lines. a comment is fine too, but don't want people uncommenting it unless they really understand what is going on (so put the comment in .travis.yml where it will be paid a bit more attention).

nparley replied:
Should I invent or find a way if doing something like:
```
if in container:
    sudo install package
```
or is it better to just remove all sudos from the travis scripts?

jreback commented on ci/install_travis.sh:
```diff
@@ -34,35 +34,37 @@ python_major_version="${TRAVIS_PYTHON_VERSION:0:1}"
+#    # probably useless, since you would need to relogin
+#    time sudo locale-gen "$LOCALE_OVERRIDE"
+#
+#    # Need to enable for locale testing. The location of the locale file(s) is
```
ok!

nparley commented:
This is now building on containers ok,  https://travis-ci.org/nparley/pandas/builds/128974759. The PR needs a clean up and a squash I think. The thing I have not looked at yet is the caching part.

jreback commented on .travis.yml:
```diff
@@ -82,13 +102,22 @@ matrix:
       - PANDAS_TESTING_MODE="deprecate"
+      addons:
+        apt:
+          packages:
```
so some of the builds are showing up above the ``allow_failures`` section. We want to keep the first 5 builds (osx, 2 x 2.7, 3.4, 3.5) as the required builds, the rest as allow_failures (mainly so things don't take so long, not because they are actually allowed to fail).

So you need to repeat anything you added on in the declaration as below; e.g. for the NUMPY_DEV build, you added:

```
addons:
  apt:
    packages:
       ...
```

just repeat this in the below section (so they are identical), then it will be fixed.

nparley replied:
I fixed this, it was caused by not indenting the new section correcting in the travis file in the allow failures section.

jreback commented:
can we print the local in ``pandas/util/print_versions.py``

```
In [2]: import locale

In [3]: locale.getlocale()
Out[3]: (None, None)
```

```
    try:
        (sysname, nodename, release,
         version, machine, processor) = platform.uname()
        blob.extend([
            ("python", "%d.%d.%d.%s.%s" % sys.version_info[:]),
            ("python-bits", struct.calcsize("P") * 8),
            ("OS", "%s" % (sysname)),
            ("OS-release", "%s" % (release)),
            # ("Version", "%s" % (version)),
            ("machine", "%s" % (machine)),
            ("processor", "%s" % (processor)),
            ("byteorder", "%s" % sys.byteorder),
            ("LC_ALL", "%s" % os.environ.get('LC_ALL', "None")),
            ("LANG", "%s" % os.environ.get('LANG', "None")),

### add here
```

I *think* this should print on the 2 non-default locales.

nparley committed:
Merge remote-tracking branch 'upstream/master' into travis

nparley committed:
Delete sudo lines from script files

nparley committed:
Fix indent bug and add comments to builds which are also in allow_failures

nparley committed:
Add LOCALE to print versions

nparley committed:
Print version info during test script

jreback commented:
@nparley looking good. is there any speed diff?

nparley committed:
Test blosc memory Backtrace

nparley commented:
I think they spin up faster but don't really build faster. Might need to investigate caching. 

There is a problem with blosc creating a stack trace. It seems to be happening in the main travis builds as well when `ci/print_version.py` is called. I have managed to get the problem in a docker container outside of travis too. I have commented out blosc version in the lastest commit to see if it goes away. The question is though does this mean blosc if miss behaving somewhere else?

jreback commented:
not sure exactly where blosc is problematic
prob a threading issue - u might have to play with it and see

you can file a separate bug report on that

nparley committed:
Revert back previous change to fix in another PR

nparley committed:
Merge remote-tracking branch 'upstream/master' into travis

jreback commented:
rebase this and ping when green.

jreback commented:
squash too

nparley committed:
Merge remote-tracking branch 'upstream/master' into travis

# Conflicts:
#	pandas/util/print_versions.py

nparley commented:
I wonder if `source activate pandas && ci/print_versions.py` should be removed from the after script. I think it better now the versions info being printed in the `ci/script.sh` file as that reflects the locale that has been set. As the after script is run in a new shell the locale does not reflect the locale that was set in `ci/script.sh`

nparley committed:
Experiment adding miniconda to cache

nparley committed:
Fix missing $

jreback commented:
I think the local setting needs to be in a separate scripts that you can call in ``scripts.sh`` AND before the post script print_versions. We need to have that separate as it makes a distinct section in travis.

you could also change the after_script to ``ci/after_print_versions.sh`` or something and put the locale, then source, then run print_versions.py there

nparley committed:
Experiment with caching

nparley committed:
Change cache name

nparley committed:
Don't recreate pandas if miniconda is from cache

nparley committed:
Bug fix

nparley committed:
Try adding ccache to travis

nparley committed:
Bug fix

nparley committed:
Try caching ccache as well

nparley committed:
Bug fix

nparley committed:
Bug fix

nparley committed:
Test ccache

nparley commented:
I have created this branch https://github.com/nparley/pandas/compare/travis...nparley:travis_cache which adds travis caching to this PR and removes the old caching. It shaves some time off the builds and almost gets the builds down below 3 hours. 

The logic for caching is:

cache `$home/miniconda` for each build

cache `$home/.cccache` for each Linux build and use the ccache that's installed on travis. 

cache `$home/.cache` for each build with the cython files. 

This is the build https://travis-ci.org/nparley/pandas/builds/134152663

jreback commented on ci/install_travis.sh:
```diff
@@ -34,55 +34,31 @@ python_major_version="${TRAVIS_PYTHON_VERSION:0:1}"
+    conda config --add channels http://conda.anaconda.org/pandas || exit 1
+    conda config --set ssl_verify false || exit 1
 
+    # Useful for debugging any issues with conda
```
I should always show ``conda info`` (whether cached or not)

jreback commented:
take out the old caching stuff as well. ``prep_cache/submit_cache``.

Further is this cached somehow by the branch-name-user? e.g. we only really want to cache pydata/pandas master (everything else should be fresh).

nparley commented:
Each brach has it's own cache. I.e. there is a master cache, then for the first branch push off master the build will use the master cache but then create it's own cache and use that for the next push. For PRs they use the cache of the branch that they will be merged into but will not write to the cache, only pushes to the branches or master do that. This means that feature branches have there own build cache separate from the master. 

As for if it should be clean off the master:

- For the miniconda bit I think that's ok if it's coming from cache. The script runs a conda update and then if any of the versions have changed in the requirements they will be updated. So I don't see any need for a fresh miniconda install. 
- The cython bit checks if any cython files have changed in the git diff of the commit and if so rebuilds the cython files, should be ok but maybe should rebuild. 
- The compiler cache should maybe not be used incase it hides a bug. 

The solution is to check the travis variables like `$TRAVIS_PULL_REQUEST` and `$TRAVIS_BRANCH` in the scripts. For example if we don't want any compiler or cython caching outside of the master I will just add `if $TRAVIS_BRANCH == 'master'` around the cache parts.

nparley committed:
Code clean up

nparley committed:
Test commit

nparley committed:
bug fix

nparley committed:
Bug fix

nparley committed:
Bug fix

jreback commented:
I am actually ok with leaving caching on for all (not sure how travis actually ages caching, but that's there issue). Though maybe add a variable CACHE=True (or something), or have a single build (in the allowed failures section), that *doesn't* cache.

nparley committed:
Touch msgpack

nparley committed:
Remove sudo even though file is not being used

nparley committed:
Remove sudo even though file is not being used

nparley committed:
Change way version is displayed

nparley committed:
Change way version is displayed

nparley committed:
Add use cache

nparley committed:
Merge branch 'travis_cache' into travis

nparley committed:
Merge remote-tracking branch 'upstream/master' into travis

nparley committed:
Grep - P is not available on mac OS X

nparley committed:
Add cython files to the cache even if we have not used the cache. Therefore keeping cache up to date for if it is turned on in the future. Add comments to the travis file about use_cache.

nparley committed:
Only on back 2 commits not 3 when looking for python file changes. Make sure cython tar file is removed if it's going to be remade.

nparley commented:
Which is the best build not to cache?

jreback commented:
the 3.5 - last of required ones -

nparley committed:
Turn off cache for 35_nslow

nparley commented:
Right I think this is mostly done. Maybe I'll keep it up to date for upstream/master for a few commits to make sure it behaves, then I'll squash it.

jreback commented on ci/submit_cython_cache.sh:
```diff
@@ -0,0 +1,12 @@
+home_dir=$(pwd)
+ls "$TRAVIS_BUILD_DIR/pandas/"
+tar cf "$CACHE_File" \
+"$TRAVIS_BUILD_DIR"/pandas/{index,algos,lib,tslib,parser,hashtable}.c \
```
there are more cython files:

```
[Sun Jun 05 09:48:23 ~/pandas]$ find pandas -name '*.pyx'
pandas/algos.pyx
pandas/hashtable.pyx
pandas/index.pyx
pandas/io/sas/saslib.pyx
pandas/lib.pyx
pandas/msgpack/_packer.pyx
pandas/msgpack/_unpacker.pyx
pandas/parser.pyx
pandas/src/generated.pyx
pandas/src/inference.pyx
pandas/src/join.pyx
pandas/src/offsets.pyx
pandas/src/period.pyx
pandas/src/properties.pyx
pandas/src/reduce.pyx
pandas/src/skiplist.pyx
pandas/src/sparse.pyx
pandas/src/testing.pyx
pandas/tslib.pyx
```

nparley replied:
I have added saslib but the follow are not being cythoned at the moment so are not creating a compiled file:

```
pandas/src/generated.pyx
pandas/src/inference.pyx
pandas/src/join.pyx
pandas/src/offsets.pyx
pandas/src/properties.pyx
pandas/src/reduce.pyx
pandas/src/skiplist.pyx
```

jreback replied:
those are all included in other cython files

issue here is to make sure this is kept up to take
when anything is added / changed

nparley replied:
I have changed the untarring so that it'll just touch the cython files it untars, therefore not needing to know what the cython files are. I guess the cython files to tar up also needs to be built without needing to know what they are.

jreback commented:
ok, we need a way to invalidate the caching. IOW say a file is changed in ``ci/*`` compared to the cache. I would turn off caching. Is it possible to do that? The main issue is that caching could last forever otherwise, and will 'work' even though we updated dependencies.

nparley commented:
If `use_cache` is not set in the travis build the job will not use the cache but will write to the cache after the build. (Maybe it would be a good idea to actually delete all the files in the cache if `use_cache` is not set as well, to be extra sure it's clean). I could unset use_cache if anything has been touched in ci/* in the last commit?

jreback commented:
@nparley yes, exactly. The issue with caches is they can get stale. So you are lured into security that things worked but they actually don't.

jreback committed:
BLD: update appveyor build to use numpy 1.11 for 3.5 build

BLD: correctly install tools.tests.data

closes #13472

ravinimmi committed:
BUG: fix to_datetime to handle int16 and int8

closes #13451

Author: Ravi Kumar Nimmi <nimmi.ravikumar@gmail.com>

Closes #13464 from ravinimmi/bugfix and squashes the following commits:

dc4944d [Ravi Kumar Nimmi] BUG: fix to_datetime to handle int16 and int8

cmazzullo committed:
BUG: df.pivot_table: margins_name ignored when aggfunc is a list

closes #13354

Author: cmazzullo <chris.mazzullo@gmail.com>

Closes #13435 from cmazzullo/pivot_table and squashes the following commits:

6017e25 [cmazzullo] Fixed up whitespace for pep8
b4b09bf [cmazzullo] Compared whole dataframes in test_pivot_table_margins_name_with_aggfunc_list
98ab8be [cmazzullo] Added doublequotes to variable names
669931f [cmazzullo] Added unit test for pivot_table margins_name when aggfunc is a list
0dcb78f [cmazzullo] Switched documentation to v0.18.2
d6a6036 [cmazzullo] BUG: df.pivot_table: margins_name is ignored when there aggfunc is list #13354

nparley committed:
Don't hard code in python files. Find them using the pyx files.

jreback commented:
hows this coming?

nparley committed:
Merge commit 'b06bc7ab10552079b5300f53ae9458b5b1583402' into travis

# Conflicts:
#	ci/install_travis.sh

nparley committed:
Add check cache to delete all cache if ci changes

nparley committed:
Bug fix

nparley committed:
Don't use ccache if not using conda cache

nparley committed:
Bug fix

nparley commented:
Caches are deleted if anything in ci/ is changed and the cython files are tarred by finding the pyx files and searching for the corresponding c or cpp file. Just testing.

nparley committed:
Merge commit '883df65b095709e2146b1f36c9697b92eb913731' into travis

nparley commented:
Currently cherry picking the last few commits to check the cache.

nparley committed:
Check whole PR for python files change if we are in a PR

nparley committed:
Merge commit '20dd17a70063387ccca5f7aca3dbefc904d99457' into travis

Commit:
DOC: specify correct kernelspec for converting notebooks (#13491)

nparley commented:
@jreback Although it's not needed to be safe I have added some code that will not use the cython cached files if a PR has changed any of the cython files. Do you think this is a good plan? For a normal branch it will not use the cython cache if the last two commits have any cython files changes. 

The other safeguard is to clear caches if anything in ci/ has changed in the last two commits. I have one more merge to be up to date with master.

jreback commented:
@jorisvandenbossche can you have a look

jreback commented on .travis.yml:
```diff
@@ -29,112 +39,194 @@ matrix:
-      - JOB_TAG=_DOC_BUILD
+        - JOB_NAME: "doc_build"
+        - FULL_DEPS=true
+        - DOC_BUILD=true # if rst files were changed, build docs in parallel with tests
```
you can remove this comment its old

jreback commented:
@nparley can you point to travis runs that:

- show using cache
- show cache force removal

nparley commented:
Something is still not quite correct in the PR which needs a bit of investigating. 

On my travis: https://travis-ci.org/nparley/pandas/jobs/138912152 is a build that used cache. The build before https://travis-ci.org/nparley/pandas/jobs/138892021 force removed the cache before building.

nparley committed:
Display output of git diff HEAD~2 --numstat | grep -E "ci/" and remove travis.yml comment

nparley committed:
Merge branch 'upstream' into travis

nparley committed:
Try fetch head in pull requests

nparley committed:
Fetch pull request head so we can look back out the last commits to the full request

nparley committed:
Merge branch 'upstream' into travis

nparley committed:
Merge branch 'upstream' into travis

nparley commented:
OK I have got the branch and pull request working correctly now. For the last commit (6c05a13): 

https://travis-ci.org/pydata/pandas/builds/139170756 - Are the PR travis builds
https://travis-ci.org/nparley/pandas/builds/139170744 - Are the branch travis builds

In `ci/check_cache.sh` the code checks to see if anything in ci has been touched in the last two commits and outputs for this commit, from branch build:

```bash
Not a PR: checking for changes in ci/ from last 2 commits
3	2	ci/check_cache.sh
2	2	ci/prep_cython_cache.sh
Files have changed in ci/ deleting all caches
```
and from PR build:
```bash
PR: checking for changes in ci/ from last 2 commits
From https://github.com/pydata/pandas
 * [new ref]         refs/pull/12946/head -> PR_HEAD
3	2	ci/check_cache.sh
2	2	ci/prep_cython_cache.sh
Files have changed in ci/ deleting all caches
```
As ci files were changed in the commit before this one the cache will be deleted and a fresh build will be used. The next commit should in theory now use the cache as long as no files in ci are hit. I push the next merge now and test this. 

(The problem was the travis checks out a detached head of the PR merge and so the `git diff head~2` was just showing the difference between master and the merge and not the changes that were made from on PR commit to next commit. I am hoping this has been fixed by using `git fetch origin pull/${TRAVIS_PULL_REQUEST}/head:PR_HEAD` to fetch the pull request into a branch and looking back at that history.)

nparley commented:
The next commit bbc7b35 has not had any files changed in ci. So we should now be able to use cache. In check cache we have:

```bash
Not a PR: checking for changes in ci/ from last 2 commits
```
and
```bash
PR: checking for changes in ci/ from last 2 commits
From https://github.com/pydata/pandas
 * [new ref]         refs/pull/12946/head -> PR_HEAD
```

I.e. nothing changed in ci. So the cache has not been deleted. However in `ci/prep_cython_cache.sh` we correctly pick up that one of the cython files has changed and therefore we should not use the cython cache.

```bash
cython_files.tar  motd.legal-displayed	pip
Cache available
Not a PR: checking for cython files changes from last 2 commits
1	3	pandas/tslib.pyx
number of cython files changed: 1
Rebuilding cythonized files
Use cache = true
Clear cache = 1
```
and
```bash
cython_files.tar  motd.legal-displayed	pip
Cache available
PR: checking for any cython file changes from last 5 commits
1	3	pandas/tslib.pyx
number of cython files changed: 1
Rebuilding cythonized files
Use cache = true
Clear cache = 1
```

In `ci/install_travis.sh` we then have:

```bash
Miniconda install already present from cache: /home/travis/miniconda
update conda
Fetching package metadata .......
# All requested packages already installed.
# packages in environment at /home/travis/miniconda:
#
...
Using ccache
gcc: /usr/lib/ccache/gcc
ccache: /usr/bin/ccache
...
ccache gcc -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC 
```

but also we can see that the cython cache is indeed not been being used:

```bash
cythoning pandas/index.pyx to pandas/index.c
```

Now we need another two commits that I can merge that don't have a cython change. To test that..

jreback commented on ci/check_cache.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
+if [ "$TRAVIS_PULL_REQUEST" == "false" ]
+then
```
clever!

jreback commented:
ok @nparley this looks really good. @jorisvandenbossche just have a look. when everyone satisfied we can merge. Then will watch for a little bit to make sure actually working.

nparley commented:
@jreback There is a test failing which is not on the master builds. The local override is set to `LOCALE_OVERRIDE="it_IT.UTF-8"` and the problem is:

`AssertionError: "Invalid argument" does not match "[Errno 22] Argomento non valido"`

It's coming from this bit of code:

```python
def test_constructor_bad_file(self):
        if is_platform_windows():
            raise nose.SkipTest("skipping construction error messages "
                                "tests on windows")

        non_file = StringIO('I am not a file')
        non_file.fileno = lambda: -1

        msg = "Invalid argument"
        tm.assertRaisesRegexp(mmap.error, msg, common.MMapWrapper, non_file)
```
Argomento non valido is Italian for Invalid argument, but the question is why this is not currently being picked up?

nparley committed:
Merge branch 'upstream' into travis

jorisvandenbossche commented:
I am not that familiar with travis details, but generally looks good. Thanks for all your work on this!

jreback commented:
ok, so @nparley can you squash everything down. ping when ready to merge.

nparley commented:
@jreback I have made another PR to fix the build error with mmap and the Italian error message here: https://github.com/pydata/pandas/pull/13507. Maybe this should be fixed before merge so all the builds will pass on merge. This branch tested with this build https://travis-ci.org/nparley/pandas/builds/139893325 is this PR merged with PR13507.

jreback commented:
thanks @nparley tremendous effort!

keep on the lookout for build the next few days and give a period check to make sure things still look ok.

thanks again!

TomAugspurger commented:
Great stuff @nparley!
